### PR TITLE
Notif haf

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Fixing the checkpatch warnings.

**Comment by jenswi-linaro** [comment]
I'm investigating the Hafnium error.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: arm64.h: add DAIFBIT_{NATIVE,FOREIGN}_INTR" and
"core: move hafnium.h into hfic.c".

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: ffa: exit with native interrupts unmasked", but may have a second view.

**Comment by jforissier** [comment]
> It looks like the upgrade to QEMU v8.1.2 included some new prerequisites.

Absolutely :wink: no worries I'll take care of that, it should not take more than a few minutes

**Comment by jenswi-linaro** [comment]
Rebased

**Comment by jenswi-linaro** [comment]
"core: ffa: add notifications with SPMC at S-EL2 or EL3" and "core: hfic: fix HF_INTERRUPT HVC calls" remain to be reviewed.

**Comment by jenswi-linaro** [comment]
Rebased, please help review "core: ffa: add notifications with SPMC at S-EL2 or EL3" and "core: hfic: fix HF_INTERRUPT HVC calls".

**Comment by jenswi-linaro** [comment]
Rebased to hopefully fix the build error.

**Comment by jenswi-linaro** [comment]
It seems that Rust is unhappy with something. @b49020 @jforissier  do you know what's wrong?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: hfic: fix HF_INTERRUPT HVC calls" and
"core: ffa: add notifications with SPMC at S-EL2 or EL3"
with comments addressed.

**Comment by b49020** [comment]
> It seems that Rust is unhappy with something. @b49020 @jforissier do you know what's wrong?

It looks like there is some Rust toolchain conflict among buildroot provided one vs the one we have installed. Will have a deeper look to resolve those conflicts.

File: core/arch/arm/kernel/thread_spmc.c [hunk 1]
```diff
@@ -1584,9 +1636,20 @@ void notif_send_async(uint32_t value)
...
+	/* global notification, delay notification interrupt */
+	uint32_t flags = BIT32(1);
+	TEE_Result res = TEE_SUCCESS;
+
+	assert(value == NOTIF_VALUE_DO_BOTTOM_HALF && spmc_notif_is_ready &&
+	       do_bottom_half_value >= 0);
+	res = ffa_set_notification(notif_vm_id, my_endpoint_id, flags,
+				   BIT64(do_bottom_half_value));
+	if (res) {
+		EMSG("notification set failed with error %d", res);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc.c:1650]: Sorry, wrong type, I"ll update to `int`.

**Comment by jenswi-linaro** [comment]
Comments addressed and tags applied.

**Comment by b49020** [comment]
> > It seems that Rust is unhappy with something. @b49020 @jforissier do you know what's wrong?
> 
> It looks like there is some Rust toolchain conflict among buildroot provided one vs the one we have installed. Will have a deeper look to resolve those conflicts.

Since this is affecting all the PRs build, I propose https://github.com/OP-TEE/build/pull/726 for now until I fix the underlying issue.

File: core/arch/arm/kernel/thread_spmc_a64.S [hunk 2]
```diff
@@ -48,8 +48,13 @@ LOCAL_FUNC ffa_msg_send_direct_resp , :
 	mov	x2, #FFA_PARAM_MBZ			/* RES MBZ */
 
 .ffa_msg_loop:
-	/* Invoke SMC with caller provided parameters */
+	/*
+	 * Native interrupts unmasked while invoking SMC with caller
+	 * provided parameters.
+	 */
+	msr	daifclr, #DAIFBIT_NATIVE_INTR
```
Comment (odeprez) [inline core/arch/arm/kernel/thread_spmc_a64.S:55]: I gather the change is merged now. Just to state that I retested by reverting this change, and using upstream Hafnium which implements a proper fix https://review.trustedfirmware.org/c/hafnium/hafnium/+/25051
On OP-TEE  exiting the SPMC while a virtual interrupt is pending, the SPMC resumes again OP-TEE using the FFA_INTERRUPT interface signaling the virtual interrupt pending.

Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_spmc_a64.S:55]: Thanks, good to know.

**Pull request merged**